crates/assists/src/handlers/extract_module_to_file.rs

```diff
@@ -0,0 +1,94 @@
+use ide_db::source_change::FileSystemEdit;
+use vfs::FileId;
+use syntax::{
+    ast::{
+        self,
+        NameOwner
+    },
+    AstNode,
+    TextSize
+};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: extract_module_to_file
+//
+// This assist extract module to file.
+//
+// ```
+// mod tests {<|>
+//     #[test] fn t() {}
+// }
+// ```
+// ->
+// ```
+// mod tests;
+//
+// #[test] fn t() {}
+// ```
+pub(crate) fn extract_module_to_file(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let assist_id = AssistId("extract_module_to_file", AssistKind::RefactorExtract);
+    let assist_label = "Extract module to file";
+    // TODO: how to get krate_name?
+    let krate_name = "foo";
+    let module = ctx.find_node_at_offset::<ast::Module>()?;
+    let module_items = module.item_list()?;
+    let module_items_text = module_items.to_string();
+    let module_name = module.name()?;
+    let target = module.syntax().text_range();
+    // TODO: how to increment FileId for new module file?
+    let new_file_id = FileId(1);
+    acc.add(assist_id, assist_label, target, |builder| {
+        let new_file_create = FileSystemEdit::CreateFile {
+            anchor: new_file_id,
+            dst: format!("/{}/{}.rs", krate_name, module_name),
+        };
+        builder.create(new_file_create);
+        // TODO: how to replace module decl?
+        builder.replace(target, format!("mod {};\n", module_name));
+        builder.edit_file(new_file_id);
+        let updated = update_module_items_string(module_items_text);
+        builder.insert(TextSize::from(0), updated)
+    })
+}
+fn update_module_items_string(items_str: String) -> String {
+    let mut items_string_lines: Vec<&str> = items_str.lines().collect();
+    items_string_lines.pop(); // Delete last line
+    items_string_lines.reverse();
+    items_string_lines.pop(); // Delete first line
+    items_string_lines.reverse();
+
+    let string = items_string_lines
+        .into_iter()
+        .map(|line| line.replacen("    ", "", 1)) // Delete indentation
```

Comment:
Maybe we should have something that just eats leading whitespace?

Reply:
module_items has method of `dedent` and is it better to use it?